instrumentation/java-http-client/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/javahttpclient/HttpClientInstrumentation.java

```diff
@@ -65,87 +67,114 @@ public void transform(TypeTransformer transformer) {
         HttpClientInstrumentation.class.getName() + "$SendAsyncAdvice");
   }
 
-  @SuppressWarnings("unused")
-  public static class SendAdvice {
+  public static class AdviceScope {
+    private final Context context;
+    private final Context parentContext;
+    private final Scope scope;
+    private final CallDepth callDepth;
+
+    private AdviceScope(Context parentContext, Context context, Scope scope, CallDepth callDepth) {
+      this.parentContext = parentContext;
+      this.context = context;
+      this.scope = scope;
+      this.callDepth = callDepth;
+    }
 
-    @Advice.OnMethodEnter(suppress = Throwable.class)
-    public static void methodEnter(
-        @Advice.Argument(value = 0) HttpRequest httpRequest,
-        @Advice.Local("otelContext") Context context,
-        @Advice.Local("otelScope") Scope scope) {
+    @Nullable
+    public static AdviceScope start(HttpRequest request) {
+      return start(request, null);
+    }
+
+    private static AdviceScope start(HttpRequest request, CallDepth callDepth) {
       Context parentContext = currentContext();
-      if (!instrumenter().shouldStart(parentContext, httpRequest)) {
-        return;
+      if (!instrumenter().shouldStart(parentContext, request)) {
+        return null;
+      }
+
+      Context context = instrumenter().start(parentContext, request);
+      return new AdviceScope(parentContext, context, context.makeCurrent(), callDepth);
+    }
+
+    public void end(HttpRequest request, HttpResponse<?> response, Throwable throwable) {
+      scope.close();
+      instrumenter().end(context, request, response, throwable);
+    }
+
+    public static AdviceScope startWithCallDepth(HttpRequest httpRequest) {
+      CallDepth callDepth = CallDepth.forClass(HttpClient.class);
+      if (callDepth.getAndIncrement() > 0) {
+        return new AdviceScope(null, null, null, callDepth);
+      }
+      return start(httpRequest, callDepth);
+    }
+
+    public boolean endWithCallDepth(HttpRequest httpRequest, Throwable throwable) {
+
+      if (callDepth.decrementAndGet() > 0) {
+        return false;
+      }
+
+      scope.close();
+      if (throwable != null) {
+        instrumenter().end(context, httpRequest, null, throwable);
+        return false;
       }
+      return true;
+    }
 
-      context = instrumenter().start(parentContext, httpRequest);
-      scope = context.makeCurrent();
+    public CompletableFuture<HttpResponse<?>> wrapFuture(
```

Comment:
Similar pattern: A reason not to store `HttpRequest` in the `AdviceScope` to avoid having to pass it in the `end`s and `wrapFuture()`?